src/libexpr/primops/fetchGit.cc

```diff
@@ -165,16 +167,35 @@ GitInfo exportGit(ref<Store> store, const std::string & uri,
         if (e.errNo != ENOENT) throw;
     }
 
-    auto source = sinkToSource([&](Sink & sink) {
-        RunOptions gitOptions("git", { "-C", cacheDir, "archive", gitInfo.rev });
-        gitOptions.standardOut = &sink;
-        runProgram2(gitOptions);
-    });
-
     Path tmpDir = createTempDir();
     AutoDelete delTmpDir(tmpDir, true);
 
-    unpackTarfile(*source, tmpDir);
+    // Submodule support can be improved by adding caching to the submodules themselves. At the moment, only the root
+    // repo is cached.
+    if (fetchSubmodules) {
+        Path tmpGitDir = createTempDir();
+        AutoDelete delTmpGitDir(tmpGitDir, true);
+
+        runProgram("git", true, { "init", tmpDir, "--separate-git-dir", tmpGitDir });
+        runProgram("git", true, { "-C", tmpDir, "fetch", "--quiet", "--force",
+                                  "--", cacheDir, fmt("%s:%s", *ref, *ref) });
+
+        runProgram("git", true, { "-C", tmpDir, "checkout", "--quiet", "FETCH_HEAD" });
+        runProgram("git", true, { "-C", tmpDir, "remote", "add", "origin", uri });
+        runProgram("git", true, { "-C", tmpDir, "submodule", "--quiet", "update", "--init" });
+
+        deletePath(tmpDir + "/.git");
```

Comment:
Submodules also leave .git _files_ in the tree, looking like this (extracted from the end of a run of ./tests/fetchGitSubmodules.sh):

```
gitdir: ../../nix-26667-1/modules/sub
```

I think the best way to handle that is to recursively delete `.git` directories and files.

Reply:
Does that mean that the tree in the Nix store contains `.git` directories? If so it's important to remove them, since we must have a canonical representation for reproducibility.

Reply:
I agree. Yes, the PR as-is has .git files in the nix store. My follow up patch [here](https://github.com/NixOS/nix/pull/3166#issuecomment-573544529) fixes that.